### PR TITLE
Hash test names instead of test suite name

### PR DESCRIPTION
**Pull request by rodrigozhou**
## What changed?

Hash test names instead of test suite name

## Why?

Try to distribute the tests across the partitions instead of the test suite.

### CODE REVIEW

File: tests/functional_test_base.go [hunk 1]
```diff
@@ -217,7 +219,7 @@ func checkTestShard(t *testing.T) {
 		t.Fatal("Couldn't convert TEST_SHARD_INDEX")
 	}
 
-	// This was determined empirically to distribute our existing test names + run times
+	// This was determined empirically to distribute our existing test names
 	// reasonably well. This can be adjusted from time to time.
 	// For parallelism 4, use 11. For 3, use 26. For 2, use 20.
 	const salt = "-salt-26"
```
Comment (dnr) [inline tests/functional_test_base.go:225]: If you're doing by sub-test you can just remove the salt and use the name directly. It was just there to force the large ones to fall in separate shards

**Review by dnr — COMMENTED** [review]
Cool, this is a nice way to do it. It does mean a little more setup/teardown overhead per shard since (basically) all suites will run in all shards. But should be fine.

Can you just add a comment to FunctionalTestBase somewhere saying that all suites that embed it must call the "parent" SetupTest from their SetupTest, otherwise it'll break the sharding?

File: tests/functional_test_base.go (hunk 1 above)
Comment (rodrigozhou) [reply tests/functional_test_base.go:225]: I tried removing, but it actually distributed the tests quite unevenly. With salt was better.

**Comment by rodrigozhou** [comment]
Sure, I'll update the PR with a comment. Also, the tests didn't seem to have executed much faster than before... I suppose it's because of the overhead of setting up the suite.

File: tests/functional_test_base.go (hunk 1 above)
Comment (dnr) [reply tests/functional_test_base.go:225]: Huh, weird. I came up with that value by collecting run times and writing a little program to search for the value that made the best distribution. Maybe we should do that again but with run times for the individual tests? It's unlikely the same value is best for suite names vs test names.

(Of course best would be to have it adjust based on historical data.. a previous startup I worked at solved this exact problem. But not worth it here.)

**Comment by dnr** [comment]
> Sure, I'll update the PR with a comment. Also, the tests didn't seem to have executed much faster than before... I suppose it's because of the overhead of setting up the suite.

Yeah :(  The overhead is why > 3 shards was hardly any improvement too. Maybe we could focus on startup/shutdown speed but I don't see much low-hanging fruit there.

File: tests/tls.go [hunk 2]
```diff
@@ -66,7 +68,9 @@ func (s *TLSFunctionalSuite) SetupTest() {
 }
 
 func (s *TLSFunctionalSuite) TearDownTest() {
-	s.sdkClient.Close()
+	if s.sdkClient != nil {
```
Comment (dnr) [inline tests/tls.go:71]: does TearDownTest get called if SetupTest does Skip()? I wouldn't have expected that

**Review by dnr — APPROVED** [review]

File: tests/tls.go (hunk 2 above)
Comment (rodrigozhou) [reply tests/tls.go:71]: It's called. The tests were panicking before this change.

**Pull request merged**